contracts/GnosisSafePersonalEdition.sol

```diff
@@ -0,0 +1,132 @@
+    /// @param v Array of signature V values sorted by owner addresses.
+    /// @param r Array of signature R values sorted by owner addresses.
+    /// @param s Array of signature S values sorted by owner addresses.
+    function payAndExecuteTransaction(
+        address to, 
+        uint256 value, 
+        bytes data, 
+        Enum.Operation operation, 
+        uint256 safeTxGas,
+        uint256 gasPrice,
+        uint8[] v, 
+        bytes32[] r, 
+        bytes32[] s
+    )
+        public
+    {
+        uint256 startGas = gasleft();
+        require(address(this).balance - value >= totalGasCosts(safeTxGas) * gasPrice);
+        checkHash(getTransactionHash(to, value, data, operation, safeTxGas, gasPrice, nonce), v, r, s);
+        // Increase nonce and execute transaction.
+        nonce++;
+        require(gasleft() - PAYMENT_GAS_COSTS >= safeTxGas);
+        if (!execute(to, value, data, operation, safeTxGas)) {
+            emit ExecutionFailed();
+        }
+        uint256 gasCosts = totalGasCosts(startGas - gasleft());
+
+        // We transfer the calculated tx costs to the tx.origin to avoid sending it to intermediate contracts that have made calls
+        // solium-disable-next-line security/no-tx-origin
+        tx.origin.transfer(gasCosts * gasPrice);
+    }
+
+    /// @dev Calculates the total gas costs for a safe transaction with the gas costs for the execution of the transaction.
+    /// @param executionGas Gas costs for the execution of the safe transaction.
+    /// @return Total gas costs for the execution (this includes gas costs for the payment to tx.origin, base transaction and payload data).
+    function totalGasCosts(uint256 executionGas) 
+        public 
+        pure
+        returns (uint256) 
+    {
+        return executionGas + PAYMENT_GAS_COSTS + BASE_TX_GAS_COSTS + msg.data.length * DATA_BYTE_GAS_COSTS;
+    }
+
+    /// @dev Allows to estimate a Safe transaction. 
+    ///      This method can only be used by the safe itself in a transaction. When estimating set `from` to the address of the safe.
+    ///      Since the `estimateGas` function includes refunds, call this method to get an estimated of the costs that are deducted from the safe with `payAndExecuteTransaction`
+    /// @param to Destination address of Safe transaction.
+    /// @param value Ether value of Safe transaction.
+    /// @param data Data payload of Safe transaction.
+    /// @param operation Operation type of Safe transaction.
+    /// @return Estimate without refunds and overhead fees (base transaction and payload data gas costs).
+    function estimate(address to, uint256 value, bytes data, Enum.Operation operation)
+        public
+        authorized
+        returns (uint256)
+    {
+        uint256 startGas = gasleft();
+        require(execute(to, value, data, operation, gasleft()));
+        return startGas - gasleft();
+    }
+
+    function checkHash(bytes32 hash, uint8[] v, bytes32[] r, bytes32[] s)
```

Comment:
Add natspec here?

Reply:
I will change it to internal, no need to expose that in the abi